File: 04 09 25/firstlastelement.cpp
#include <iostream>
#include <vector>
using namespace std;

class Solution {
public:
    vector<int> searchRange(vector<int>& nums, int target) {
        int first = findBound(nums, target, true);
        if (first == -1) return {-1, -1};
        int last = findBound(nums, target, false);
        return {first, last};
    }

private:
    int findBound(vector<int>& nums, int target, bool isFirst) {
        int left = 0, right = nums.size() - 1;
        int bound = -1;

        while (left <= right) {
            int mid = left + (right - left) / 2;
            if (nums[mid] == target) {
                bound = mid;
                if (isFirst) {
                    right = mid - 1;  // search left half
                } else {
                    left = mid + 1;   // search right half
                }
            } else if (nums[mid] < target) {
                left = mid + 1;
            } else {
                right = mid - 1;
            }
        }

        return bound;
    }
};

int main() {
    Solution sol;
    vector<int> nums = {5,7,7,8,8,10};
    vector<int> res = sol.searchRange(nums, 8);
    cout << "[" << res[0] << ", " << res[1] << "]\n"; // Output: [3, 4]

    res = sol.searchRange(nums, 6);
    cout << "[" << res[0] << ", " << res[1] << "]\n"; // Output: [-1, -1]
}

File: 04 09 25/firstlastelement.rb
class Solution
  def search_range(nums, target)
    first = find_bound(nums, target, true)
    return [-1, -1] if first == -1
    last = find_bound(nums, target, false)
    [first, last]
  end

  private

  def find_bound(nums, target, is_first)
    left = 0
    right = nums.length - 1
    bound = -1

    while left <= right
      mid = (left + right) / 2
      if nums[mid] == target
        bound = mid
        if is_first
          right = mid - 1  # search left half
        else
          left = mid + 1   # search right half
        end
      elsif nums[mid] < target
        left = mid + 1
      else
        right = mid - 1
      end
    end

    bound
  end
end
